### PR TITLE
Add to node parsing capabilities

### PR DESCRIPTION
**Pull request by shageman**
## What are you trying to accomplish?
In the codebase I am working on I see several errors related to edge cases where the code parsing fails. I am working to make packwerk run.

## What approach did you choose and why?
I added a bunch of defensive statements that makes packwerk continue to run even when individual nodes fail to parse.

## What should reviewers focus on?
This doesn't seem like a cohesive PR at the moment. Although the tests do pass, how should I break these changes up so merging would be considered?

## Type of Change

- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Code refactor (non-breaking change that doesn't add functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Checklist

- [X] It is safe to simply rollback this change.

### CODE REVIEW

**Comment by shageman** [comment]
How do I complete the CLA step?

**Comment by wildmaples** [comment]
> How do I complete the CLA step?

@shageman https://cla.shopify.com/

File: lib/packwerk/parsed_constant_definitions.rb
Comment (doodzik) [inline lib/packwerk/parsed_constant_definitions.rb:41]: From what I can tell, you are only adding a rescue block here, correct?
In that case, I would suggest getting rid of the `begin` statement in this method. 
This is a good example of what I mean: http://seejohncode.com/2012/04/17/short-tip-rescuing-a-method-in-ruby/

File: lib/packwerk/node.rb
Comment (doodzik) [inline lib/packwerk/node.rb:36]: I don't think we want to rescue against a TypeError in this method, because we expect it to raise a TypeError as seen on line 38. I would suggest rescuing the type of error in the method that calls this method.

**Review by doodzik — COMMENTED** [review]
Your changes look like a great addition 👍 I left two small comments but other than that it looks good.

File: lib/packwerk/parsed_constant_definitions.rb
Comment (shageman) [reply lib/packwerk/parsed_constant_definitions.rb]: Yes!

File: lib/packwerk/node.rb
Comment (shageman) [reply lib/packwerk/node.rb]: Makes sense!

**Comment by shageman** [comment]
@doodzik I believe I addressed your comments with the new changes and hope to have also appeased the linter.

**Comment by shageman** [comment]
Scratch that, my changes were too aggressive that broke that I need to go back and run tests locally. I hope to have another version tomorrow

File: lib/packwerk/const_node_inspector.rb
Comment (doodzik) [inline lib/packwerk/const_node_inspector.rb:30]: This PR looks good to me, but I would like to get the approval of @tomstuart because he raised an issue [about defensive node parsing on one of my open PRs](https://github.com/Shopify/packwerk/pull/39/files#r513344732). From my understanding, this gem's design focuses on doing a good enough job uncovering dependency/privacy violations. In my opinion, defensive node parsing would fall under that because we don't fail if packwerk cannot handle the code. The downside is that we don't learn about parsing issues.

Comment (shageman) [reply lib/packwerk/const_node_inspector.rb:30]: I totally agree. I would love to rather propose a fix for the parsing errors we are running into, but am not deep enough into packwerk and parser to understand much of what's going on. 

I would be happy to pair on this or run non-defensive versions against the codebase I am looking at to give more input into what might be going on...

**Comment by shageman** [comment]
If I fix the linting error as proposed by the check, all tests fail: https://github.com/Shopify/packwerk/pull/54/checks?check_run_id=1328856280

**Comment by exterm** [comment]
> all tests fail

Could you link to the build with the failing tests? the one you link to only has the failing linter.

**Comment by shageman** [comment]
> > all tests fail
> 
> Could you link to the build with the failing tests? the one you link to only has the failing linter.

Yes - me too 🤔

**Comment by shageman** [comment]
@exterm @doodzik I took another stab at the changes needed to make it work for our codebase... this time a bit more focussed and with tests.

File: lib/packwerk/parsed_constant_definitions.rb
Comment (exterm) [inline lib/packwerk/parsed_constant_definitions.rb:55]: I'm a bit concerned about this part. It looks like in the future, this could lead to parsing bugs being silently ignored, and the behavior of packwerk regressing.

From your tests, it looks like the case that you want to fix is dynamic namespacing (`self::SOME_CONSTANT`). The fact that packwerk chokes on that seems like a bug to me. Can we find a solution that fixes this exact bug, instead of introducing a clause that ignores all unknown nodes?

Comment (shageman) [reply lib/packwerk/parsed_constant_definitions.rb]: I have removed the offending lines you point to and added the ability for packwerk to parse the `self::` syntax

File: lib/packwerk/node.rb
Comment (exterm) [inline lib/packwerk/node.rb:198]: Do you still need this now that you are handling `self::` explicitly? The only test that fails when I revert this change is the test that explicitly tests this change.

File: test/unit/reference_extractor_test.rb
Comment (exterm) [inline test/unit/reference_extractor_test.rb:53]: I guess you can make this test description more specific now, something about namespacing by `self`?

**Comment by exterm** [comment]
Thank you for the adjustments! This PR is starting to look like a very valuable bugfix.

**Review by doodzik — APPROVED** [review]
The PR is looking good! Once Philips comment is addressed and fix the merge conflict we can ship this PR 🚢

File: lib/packwerk/node.rb
Comment (shageman) [reply lib/packwerk/node.rb]: ok!

File: test/unit/reference_extractor_test.rb
Comment (shageman) [reply test/unit/reference_extractor_test.rb]: good catch! Thanks!

**Comment by shageman** [comment]
Thanks, @doodzik - I have addressed all open comments

**Comment by exterm** [comment]
looking good! I'm going to test this on our core monolith.

**Comment by exterm** [comment]
It's complicated by the fact that packwerk main is currently broken on our monolith. We may want to fix that first before merging this. Sorry for the delay Stephan.

**Comment by exterm** [comment]
We fixed the main branch. @shageman , can you rebase? I'll manually test your branch on our monolith afterwards (I would have rebased myself, but there are merge conflicts and I sadly don't have the time to look into them right now)

**Comment by shageman** [comment]
@exterm done. Thanks!

**Comment by exterm** [comment]
Ran it on the core monolith without problems.

**Review by exterm — APPROVED** [review]
Thank you for your contribution Stephan!

**Pull request merged**
